File: backend/src/routers/tasks.py
from fastapi import APIRouter
from src.schemas.task import Task, TaskCreate, TaskInBD, TaskUpdate
from src.dependencies.auth import authDep_user, authDep_admin
from src.dependencies.db import sessionDep
from fastapi_cache.decorator import cache
from src.services.tasks import TaskService


router = APIRouter(prefix="/tasks", tags=["Tasks"])
expire_time = 360


@router.post("/create_task", summary="Создать задачу")
async def create_task_router(
    task_data: TaskCreate, db: sessionDep, user: authDep_user
) -> Task:
    created_task = await TaskService.create_task(task_data, db, user)
    return created_task


@router.get("/get_task_by_id/{task_id}", summary="Получить задачу по id")
@cache(expire=expire_time)
async def get_task_by_id_router(task_id: int, db: sessionDep) -> TaskInBD:
    task = await TaskService.get_task_by_id(task_id, db)
    return task


@router.get("/get_all_tasks", summary="Получить все задачи")
@cache(expire=expire_time)
async def get_all_tasks_router(db: sessionDep) -> dict[str, list[Task]]:
    all_tasks = await TaskService.get_all_tasks(db)
    return all_tasks


@router.get("/get_user_tasks/{user_id}", summary="Получить все задачи у пользователя")
@cache(expire=expire_time)
async def get_user_tasks_router(user_id: int, db: sessionDep) -> dict[str, list[Task]]:
    user_tasks = await TaskService.get_user_tasks(user_id, db)

    return user_tasks


@router.delete("/delete_task/{task_id}", summary="Удалить задачу по id")
async def delete_task_router(task_id: int, db: sessionDep, user: authDep_user) -> Task:
    deleted_task = await TaskService.delete_task_by_id(task_id, db)
    return deleted_task


@router.put("/update_task/{task_id}", summary="Обновить задачу по id")
async def update_task_router(
    task_id: int, new_data: TaskUpdate, db: sessionDep
) -> Task:
    updated_task = await TaskService.update_task_by_id(task_id, new_data, db)
    return updated_task


@router.post("/reset_tasks", summary="Сбросить все задачи")
async def reset_tasks_router(db: sessionDep, user: authDep_admin) -> dict[str, str]:
    await TaskService.reset_tasks(db)
    return {"status": "Tasks was successfully reset"}

File: frontend/api/tasks.js
import { operateData } from './methods.js'

export async function getTasks() {
    try {
        let url = 'http://127.0.0.1:8000/api/users/me'
        const response = await operateData({
            url: url,
            method: 'GET',
            isProtected: true
        })
        const usersData = response.ok ? await response.json() : []
        
        return usersData ? usersData?.tasks : []
    } catch (error) {
        console.error('getTasks error:', error);   
    }
}

export async function addTask(params = {}) {
    try {
        const url = 'http://127.0.0.1:8000/api/tasks/create_task'
        const createdTask = await operateData({
            url: url,
            method: 'POST',
            data: params,
            isProtected: true
        })
        return createdTask
    } catch (error) {
        console.error('addTask error:', error);
    }
}

export async function deleteTask(taskId) {
    try {
        const url = `http://127.0.0.1:8000/api/tasks/delete_task/${+taskId}`
        const deletedTask = await operateData({
            url: url,
            method: 'DELETE',
            isProtected: true
        })
        return deletedTask
    } catch (error) {
        console.error('deleteTask error:', error);
    }
}

export async function updateTask(taskId, params) {
    try {
        const url = `http://127.0.0.1:8000/api/tasks/update_task/${taskId}`
        const updatedTask = await operateData({
            url: url,
            method: 'POST',
            data: params,
            isProtected: true
        })
        return updatedTask
    } catch (error) {
        console.error('updateTask error:', error);
    }
}
